Fix bounds for metadata layers

sssangha opened the pull request:
Found bug in way lat/lon metadata arrays were initiated. Specifically, incorrect imposed dimensions translated into an apparent offset over overlapping regions when frames were extracted separately as so:
![Screenshot 2023-05-16 at 2 43 42 AM](https://github.com/aria-tools/ARIA-tools/assets/13227405/7508479a-8463-4fd2-8dc2-52de03abf243)


When the dimensions are appropriately set [here](https://github.com/aria-tools/ARIA-tools/blob/sss_metafix/tools/ARIAtools/extractProduct.py#L1265-L1272), this inconsistency goes away:
![Screenshot 2023-05-16 at 2 44 01 AM](https://github.com/aria-tools/ARIA-tools/assets/13227405/3b113fce-8542-4c1c-9faa-d17d40a73257)


And as a sanity check, this artifact is NOT inherited from the GUNWs, which show seamless overlap:
![Screenshot 2023-05-16 at 2 43 05 AM](https://github.com/aria-tools/ARIA-tools/assets/13227405/4d91381c-0859-4d7d-aca7-8936c54089db)

Commit:
Fix bounds for metadata layers

Commit:
Initiate width and length tracker

Commit:
Revert TAG commits

Commit:
remove rempt files from sequential stitcher

Commit:
Bug-fix for metadata bounds

Commit:
Specify spacing + TAP

Commit:
Remove bug in dimension check

Commit:
Fix bug in NLCD mask generator and remove temp files

sssangha commented:
My tests suggests when interpolating metadata layers, we obtain more precise results when specifying the pixel spacing + `target aligned pixel` option to `gdalWarp` (as opposed to status quo of specifying output raster dimensions).

Specifically, I compared the overlap area between overlapping frames that had been extracted in separate runs.

Before introducing these changes, there is an ~0.001 order discrepancy in the overlap region:
![Screenshot 2023-05-30 at 6 23 21 PM](https://github.com/aria-tools/ARIA-tools/assets/13227405/7cf4d6d4-2761-474a-b90d-5c487ed7011c)


When instead passing the pixel spacing + `target aligned pixel` instead, this discrepancy is on the order of ~<0.0001:
![Screenshot 2023-05-30 at 6 49 52 PM](https://github.com/aria-tools/ARIA-tools/assets/13227405/334b7ea5-3972-431d-86cf-2bafd44e93ad)

sssangha committed:
Initiate variables to track dimensions at top of function

mgovorcin commented on tools/ARIAtools/extractProduct.py:
```diff
@@ -570,21 +571,31 @@ def merged_productbbox(metadata_dict, product_dict, workdir='./',
+    arrres = [abs(arrres.GetGeoTransform()[1]),
+              abs(arrres.GetGeoTransform()[-1])]
+    ds = gdal.Warp('',
+                   gdal.BuildVRT('', product_dict[0]['unwrappedPhase'][0]),
```
would be good to define the common parameters of gdal.Warp and gdal.BuildVrt beforehand and submit it as kwargs. This could lead in little code clean up.

gdal_warp_kwargs = dict(format='MEM', targetAlignedPixels=True, multithread=True, options=['NUM_THREADS=%s'%(num_threads)]) 

gdal.Warp(**gdal_warp_kwargs)

sssangha replied:
@mgovorcin ok I adjusted all of the gdalwarp option calls in `extractProduct.py`.

There are other scripts/functions I will extend this to, but I think it's more appropriate for a follow-up PR since this one is getting a bit large! Let me know what you think

sssangha committed:
Appropriately derive lat/lon bounds from DEM

Commit:
Add gdalwarp kwarg dict

sssangha commented:
I fixed a bug in how the Lat/Lon arrays derived from the DEM were initiated. Specifically, I adjusted the stop increment in [lat](https://github.com/aria-tools/ARIA-tools/blob/sss_metafix/tools/ARIAtools/extractProduct.py#L434) and [lon](https://github.com/aria-tools/ARIA-tools/blob/sss_metafix/tools/ARIAtools/extractProduct.py#L436) as so, respectively.

These Lat/Lon arrays are used for the interpolation/intersection of the metadata layers. So with the bug, these very small shifts (here SET fields displayed in radians, with shifts translating to 0.01 mm) between overlapping frames (here `*_f2` and `*_f1`, with the stitched `*_merged` product for reference):
![Screenshot 2023-06-07 at 5 22 06 PM](https://github.com/aria-tools/ARIA-tools/assets/13227405/3636b844-ed77-4606-8bc7-569f5a064183)


After the bug-fix this apparent shift has now been (finally!) completely eliminated (note the y-axis):
![Screenshot 2023-06-07 at 5 16 15 PM](https://github.com/aria-tools/ARIA-tools/assets/13227405/7d4db080-4894-43ab-95b3-3c9e4d8775fb)

rzinke commented:
Thanks, @sssangha. Testing now.

Commit:
Fix check for dims in external correction layers

sssangha commented:
@rzinke let me know if we have the green-light to merge. Thanks!

Commit:
Fix bug in tropo dim check

Commit:
Fix bugs with mix of tropo models

rzinke commented:
Hi @sssangha. Apologies for the slow response, I was away at a conference all last week and just back in the office today...
I am still getting an error:

```
Generating: incidenceAngle - [==================================================] 20141113_20141020
Generating: lookAngle - [==================================================] 20141113_20141020
Generating: azimuthAngle - [==================================================] 20141113_20141020WARNING: azimuthAngle layer for IFG 20141113_20141020 has R² of 0.0000 and standard error of 0.4806°, automated correction applied


Extracting perpendicular baseline grids for each interferogram pair
Generating: bPerpendicular - [>                                                 ]Traceback (most recent call last):
  File "/home/rzinke/miniconda3/envs/sssARIA-tools/bin/ariaTSsetup.py", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/rzinke/tools/ARIA-tools/tools/bin/ariaTSsetup.py", line 18, in <module>
    main(inps)
  File "/home/rzinke/tools/ARIA-tools/tools/ARIAtools/tsSetup.py", line 502, in main
    prod_arr_record = export_products(standardproduct_info.products[1],
  File "/home/rzinke/tools/ARIA-tools/tools/ARIAtools/extractProduct.py", line 1104, in export_products
    finalize_metadata(outname, bounds, dem_bounds,
  File "/home/rzinke/tools/ARIA-tools/tools/ARIAtools/extractProduct.py", line 1276, in finalize_metadata
    data_array = metadata_qualitycheck( \
  File "/home/rzinke/tools/ARIA-tools/tools/ARIAtools/extractProduct.py", line 169, in __init__
    self.__run__()
  File "/home/rzinke/tools/ARIA-tools/tools/ARIAtools/extractProduct.py", line 264, in __run__
    rsquaredarr_rng, std_errarr_rng = self.__getCovar__('range')
  File "/home/rzinke/tools/ARIA-tools/tools/ARIAtools/extractProduct.py", line 221, in __getCovar__
    if j==range(0, len(mid_line.tolist()), chunk_size)[-2] and \
IndexError: range object index out of range
```

sssangha commented:
@rzinke strange.. Could you perhaps provide me with the explicit commandline run you've used to help me debug.. Also ideally does this involve a smaller subset? Thanks

bbuzzanga commented:
Additionally, @sssangha at least for a single GUNW with a single weather model, I get this error:

```Generating: troposphereWet_HRES troposphereTotal - [==================================================] 20210817_20210805
Traceback (most recent call last):
  File "/u/leffe-data2/buzzanga/Miniconda3/envs/ARIA/bin/ariaTSsetup.py", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/u/leffe-data2/buzzanga/Software_InSAR/ARIA-tools_git/tools/bin/ariaTSsetup.py", line 18, in <module>
    main(inps)
  File "/u/leffe-data2/buzzanga/Software_InSAR/ARIA-tools_git/tools/ARIAtools/tsSetup.py", line 530, in main
    dim_check(ref_arr_record, prod_arr_record)
  File "/u/leffe-data2/buzzanga/Software_InSAR/ARIA-tools_git/tools/ARIAtools/vrtmanager.py", line 569, in dim_check
    raise Exception('Inconsistent product dims '
Exception: Inconsistent product dims between products /u/leffe-data2/buzzanga/data/VLM/RAiDER_Tests/HR/HRES/troposphereTotal/HRES/20210817_20210805 and /u/leffe-data2/buzzanga/data/VLM/RAiDER_Tests/HR/HRES/coherence/20210817_20210805: check respective width (2239, None) and height (3811, None) and geotrans ((-78.5249685899999, 0.000833332999999999, -0.0, 37.6433182760001, -0.0, -0.000833333000000001), None)
```
Command is : `ariaTSsetup.py -f ./products/S1-GUNW-A-R-004-tops-20210817_20210805-230647-00079W_00036N-PP-d06e-v2_0_5.nc -m download -d download -l 'unwrappedPhase,troposphereTotal' -tm HRES -w HRES`
ariaTSsetup.py -f ./products/S1-GUNW-D-R-071-tops-20200130_20200112-135156-34956N_32979N-PP-dd0c-v2_0_4.nc -m download -d download -l 'unwrappedPhase,troposphereTotal' -tm HRES -w HRES

I used the same command except `ariaExtract.py` on dev and this error does not occur

Commit:
Fix dim check for tropo

rzinke commented:
> @rzinke strange.. Could you perhaps provide me with the explicit commandline run you've used to help me debug.. Also ideally does this involve a smaller subset? Thanks

Hi @sssangha, I reran the exact same command over the weekend and it finished without error. I was running several other processing on the machine when the previous error occurred to me, so perhaps that had something to do with it?

Commit:
Pass metadata as single precision

sssangha commented:
> 

@rzinke sounds good. Yeah I was able to grab the same pairs in question over track 41 and couldn't replicate it.

I also noticed that the traceback of the error log you reported denoted line numbers that seemed out-of-date vis-a-vis the current revision. So with that all said, since we can't replicate the error and all else looks stable, do we have the green-light to merge?

rzinke commented:
Yes! Thanks for checking.
